proposals/gc/Overview.md

```diff
@@ -1 +1,716 @@
+}
+```
+
+```
+(type $func-f64-f64 (func (param $env $anyref) (param $y f64) (result f64)))
+(type $clos-f64-f64 (struct (field $code (ref $func-f64-f64)) (field $env anyref)))
+(type $inner-env (struct (field $x i32) (field $a i32)))
+
+(func $outer (param $x f64) (result (ref $clos-f64-f64))
+  (ref_func $inner)
+  (get_local $x)
+  (f64.add (get_local $x) (f64.const 1))
+  (new $inner-env)
+  (new $clos-f64-f64)
+)
+
+(func $inner (param $anyenv anyref) (param $y f64) (result f64)
+  (local $env (ref $inner-env))
+  (block $fail (result (ref anyref))
+    (set_local $env (cast_down anyref (ref $inner-env) $fail (get_local _$anyenv)))
+    (get_local $y)
+    (get_field $inner-env $a (get_local $env))
+    (f64.add)
+    (get_field $inner-env $x (get_local $env))
+    (f64.add)
+    (return)
+  )
+  (unreachable)
+)
+
+(func $caller (result f64)
+  (local $clos (ref $clos-f64-f64))
+  (set_local $clos (call $outer (f64.const 1)))
+  (call_ref
+    (get_field $clos-f64-f64 $env (get_local $clos))
+    (f64.const 2)
+    (get_field $clos-f64-f64 $code (get_local $clos))
+  )
+)
+```
+
+Needs:
+* function pointers
+* universal type of references
+* down casts
+
+The down cast for the closure environment is necessary because expressing its static type faithfully would require first-class generics and existential types.
+
+An alternative is to provide [primitive support](#closures) for closures.
+
+
+### Parametric Polymorphism
+
+TODO: via type `anyref` and `intref`
+
+
+### Type Export/Import
+
+* Want to allow type definitions to be imported from other modules.
+* As much as possible of the above constructions should be allowed with abstract types.
+* More complicated linking patterns might require user-space linking hooks.
+* Possibly: allow abstract type exports (encapsulation)?
+* Lots of tricky details here, mostly ignore for now...
+
+
+## Basic Functionality: Simple Aggregates
+
+* Extend the Wasm type section with new constructors for aggregate types.
+* Extend the value types with new constructors for references and interior references.
+* Aggregate types are not value types, only references to them are.
+* References are never null; nullable reference types are separate.
+
+
+### Structures
+
+*Structure* types define aggregates with heterogeneous fields that are _statically indexed_:
+```
+(type $time (struct (field i32) (field f64)))
+(type $point (struct (field $x f64) (field $y f64) (field $z f64)))
+```
+Such types can be used by forming *reference types*, which are a new type of value type:
+```
+(local $var (ref $point))
+```
+
+Fields are *accessed* with generic load/store instructions that take a reference to a structure:
+```
+(func $f (param $p (ref $point))
+  (store_field $point $y (get_local $p)
+    (load_field $point $x (get_local $p))
+  )
+)
+```
+All accesses are type-checked at validation time.
+
+Structures are [allocated](#allocation) with `new` instructions that take initialization values for each field.
```

Comment:
Removed xref. Yeah, probably not worth going into details about possible optimisations at this level. As usual, implementations can do whatever is semantically equivalent. Additional language should only be necessary if we explicitly want to _forbid_ certain optimisations, e.g., like we did with TCO.